hikari/api/interaction_server.py

```diff
@@ -0,0 +1,202 @@
+    `hikari.api.special_endpoints.InteractionResponseBuilder` see
+    `hikari.impl.special_endpoints`.
+"""
+
+
+class Response(typing.Protocol):
+    """Protocol of the data returned by `InteractionServer.on_interaction`.
+
+    This is used to instruct lower-level REST server logic on how it should
+    respond.
+    """
+
+    __slots__: typing.Sequence[str] = ()
+
+    @property
+    def headers(self) -> typing.Optional[typing.Mapping[str, str]]:
+        """Headers that should be added to the response if applicable.
+
+        Returns
+        -------
+        typing.Optional[typing.Mapping[builtins.str, builtins.str]]
+            A mapping of string header names to string header values that should
+            be included in the response if applicable else `builtins.None`.
+        """
+        raise NotImplementedError
+
+    @property
+    def payload(self) -> typing.Optional[bytes]:
+        """Payload to provide in the response.
+
+        !!! note
+            If this is not `builtins.None` then an appropriate `"Content-Type"`
+            header should be declared in `Response.headers`
+
+        Returns
+        -------
+        typing.Optional[builtins.bytes]
+            The bytes payload to respond with if applicable else `builtins.None`.
+        """
+        raise NotImplementedError
+
+    @property
+    def status_code(self) -> int:
+        """Status code that should be used to respond.
+
+        Returns
+        -------
+        builtins.int
+            The response code to use for the response. This should be a valid
+            HTTP status code, for more information see
+            https://developer.mozilla.org/en-US/docs/Web/HTTP/Status.
+        """
+        raise NotImplementedError
+
+
+class InteractionServer(abc.ABC):
+    """Interface for an implementation of a Interactions compatible REST server."""
+
+    __slots__: typing.Sequence[str] = ()
+
+    @abc.abstractmethod
+    async def on_interaction(self, body: bytes, signature: bytes, timestamp: bytes) -> Response:
+        """Handle an interaction received from Discord as a REST server.
+
+        Parameters
+        ----------
+        body : builtins.bytes
+            The interaction payload.
+        signature : builtins.bytes
+            Value of the `"X-Signature-Ed25519"` header used to verify the body.
+        timestamp : builtins.bytes
+            Value of the `"X-Signature-Timestamp"` header used to verify the body.
+
+        Returns
+        -------
+        Response
+            Instructions on how the REST server calling this should respond to
+            the interaction request.
+        """
+
+    # # This overload cannot be included until component interactions are implemented
+    # @typing.overload
+    # def get_listener(
+    #     self, interaction_type: typing.Type[commands.CommandInteraction], /
+    # ) -> typing.Optional[ListenerT[commands.CommandInteraction, _MessageResponseBuilderT]]:
+    #     ...
```

Comment:
Because of stuff like this, should we merge this into a feature branch, push all the documentation updates as well as the components there and then merge to master?

Reply:
Seems kinda unnecessary to me tbh (the component stuff can just be PR'ed into master as soon as this is merged anyway) plus that'd merge even more changes into what is already going to be a giant single commit which we'd probably want to avoid doing any more of lol.

Reply:
I don't mind having huge branch merges as long as we break them down into smaller commits, which will make sure we have been able to review all the changes and make sure they will fit nicely into master, which I'm not so sure from this.

Reply:
yeah but all PR merges are squashed so a feature branch would just lead to all the commits to it being squashed into one